io-engine/src/bdev/ftl.rs

```diff
@@ -0,0 +1,258 @@
+//!
+//! The ftl bdev creates a CSAL-FTL disk.
+use std::{
+    collections::HashMap,
+    convert::TryFrom,
+    fmt::{Debug, Formatter},
+};
+
+use core::ffi::c_void;
+
+use async_trait::async_trait;
+use futures::channel::oneshot;
+use log::info;
+use nix::errno::Errno;
+use snafu::{OptionExt, ResultExt};
+use std::mem;
+use url::Url;
+
+use spdk_rs::ffihelper::errno_result_from_i32;
+use spdk_rs::{
+    libspdk::{
+        bdev_ftl_create_bdev, bdev_ftl_delete_bdev, ftl_bdev_info,
+        spdk_ftl_conf, spdk_ftl_get_default_conf,
+    },
+    UntypedBdev,
+};
+
+use crate::{
+    bdev::{dev::reject_unknown_parameters, util::uri, CreateDestroy, GetName},
+    bdev_api::{self, bdev_create, bdev_destroy, BdevError},
+    core::VerboseError,
+    ffihelper::{cb_arg, done_errno_cb, ErrnoResult},
+};
+
+pub struct Ftl {
+    /// the name of the ftl-bdev we created
+    name: String,
+    /// alias which can be used to open the bdev
+    alias: String,
+    /// uuid of the spdk bdev
+    uuid: Option<uuid::Uuid>,
+    // ftl's base bdev
+    bbdev_uri: String,
+    // ftl's cache bdev (must support VSS DIX mode)
+    cbdev_uri: String,
+}
+
+impl Debug for Ftl {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "Ftl '{}' (bbdev_uri {} cbdev_uri {})",
+            self.name, self.bbdev_uri, self.cbdev_uri
+        )
+    }
+}
+
+impl TryFrom<&Url> for Ftl {
+    type Error = BdevError;
+
+    fn try_from(uri: &Url) -> Result<Self, Self::Error> {
+        let segments = uri::segments(uri);
+        if segments.is_empty() {
+            return Err(BdevError::InvalidUri {
+                uri: uri.to_string(),
+                message: "empty path".to_string(),
+            });
+        }
+
+        let mut parameters: HashMap<String, String> =
+            uri.query_pairs().into_owned().collect();
+
+        let uuid = uri::uuid(parameters.remove("uuid")).context(
+            bdev_api::UuidParamParseFailed {
+                uri: uri.to_string(),
+            },
+        )?;
+
+        let bbdev_uri = parameters
+            .remove("bbdev")
+            .context(bdev_api::InvalidUri {
+                uri: uri.to_string(),
+                message: String::from("No bbdev parameter found"),
+            })?
+            .to_string()
+            .replace("!", "?")
+            .replace("|", "&");
+
+        let cbdev_uri = parameters
+            .remove("cbdev")
+            .context(bdev_api::InvalidUri {
+                uri: uri.to_string(),
+                message: String::from("No cbdev parameter found"),
+            })?
+            .to_string()
+            .replace("!", "?")
+            .replace("|", "&");
```

Comment:
This is because we can't have those in the "sub-uris" ? Would be good to document that at the top of the file as well.
The nx and lvs probably have the same issue, maybe I never tried using query parameters there
Can around this as well by using percent encoding?

Reply:
Yes, it is because we can't have the `sub-uris`. I have to try the percent encoding. Lets leave this conversation as unresolved for now :)

Reply:
The percent-encoding should be possible. However, the user specified URIs will get very convoluted:
See the very simple example (without additional options in the nested URIs) with my proposed encoding vs percent-encoding:
```
ftl:///ftl-1?bbdev=pcie:///0000:01:00.0&cbdev=pcie:///0000:02:00.0
```
vs.
```
ftl:///ftl-1?bbdev=pcie%3A%2F%2F%2F0000%3A01%3A00.0&cbdev=pcie%3A%2F%2F%2F0000%3A02%3A00.0
```

What do you think how we should proceed?

Reply:
Can we use a mix, example, only use percent decoding for the ? and & ?

Reply:
Yes this works. I rewrote this parsing section and now we can use partial percent-encoding for `?` and `&` . Optionally you can also use percent-encoding for the entire sub-uri.
Thanks for this suggestion :)

Reply:
Awesome :)